### PR TITLE
BUG: GH15429 transform result of timedelta from datetime

### LINKED ISSUE #10972: BUG: output of a transform is cast to dtype of input
xref #11444, #13046 for addtl tests

``` python
In [27]: df = pd.DataFrame({'a': np.random.randint(0, 5, 365), 'b': pd.date_range('2015-01-01', periods=365, freq='D')})

In [28]: df.head()
Out[28]:
   a          b
0  3 2015-01-01
1  3 2015-01-02
2  4 2015-01-03
3  2 2015-01-04
4  4 2015-01-05

In [29]: df.groupby('a').b.transform(lambda x: x.dt.dayofweek - x.dt.dayofweek.mean()).head()
Out[29]:
0   1970-01-01 00:00:00.000000000
1   1970-01-01 00:00:00.000000001
2   1970-01-01 00:00:00.000000001
3   1970-01-01 00:00:00.000000002
4   1969-12-31 23:59:59.999999997
Name: b, dtype: datetime64[ns]
```

I expected a float. No idea how difficult this will be so I marked it for 0.18. I won't have time to get to it any earlier, but if someone else wants to...

### LINKED ISSUE #15429: BUG: groupby.transform ValueError: Could not convert object to NumPy datetime
#### Problem description

As can be seen [here](http://stackoverflow.com/questions/42260764/pandas-timestamp-difference-in-groupby-transform), the error `ValueError: Could not convert object to NumPy datetime` is being thrown by:

```python
df = DataFrame(dict(A=Timestamp('20130101'), B=np.arange(5)))
result = df.groupby('A')['A'].transform(lambda x: x.max() - x.min())
```

NOTE: I will be submitting a PR shortly.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 32
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little

pandas: 0.19.2
nose: 1.3.7
pip: 8.1.2
setuptools: 18.2
Cython: None
numpy: 1.12.0
dateutil: 2.6.0
pytz: 2016.6.1
</details>

### PR DESCRIPTION
**Pull request by stephenrauch**
The transform() operation needs to return a like-indexed. To facilitate this, transform starts with a copy of the original series. Then, after the computation for each group, sets the appropriate elements of the copied series equal to the result. At that point is does a type comparison, and discovers that the timedelta is not cast-able to a datetime.

 - [x] closes #10972
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
instead of this, use ``pandas.type.cast._find_common_type`` rather than ``np.common_type``; that is some old code.

**Comment by jreback** [comment]
your bug report is the same as this: https://github.com/pandas-dev/pandas/issues/10972

so please add these as tests cases.

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -579,6 +579,8 @@ Bug Fixes
 
 
 
+- Bug in ``groupby.transform`` where calculating a ``timedelta`` from a ``datetime`` caused a ``ValueError`` (:issue:`15429`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: add 10972 as well (you can leave the other issue as well)

**Comment by jreback** [comment]
can you update

**Comment by stephenrauch** [comment]
I will take care of this this weekend...

**Comment by stephenrauch** [comment]
I tried to use `pandas.type.cast._find_common_type` rather than `np.common_type` in the function where this work was done, but one of the other test cases broke.  I looked at it, but didn't undertsand it well enough to know how to fix.

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -624,7 +624,8 @@ Bug Fixes
 - Bug in ``pd.concat()`` in which concatting with an empty dataframe with ``join='inner'`` was being improperly handled (:issue:`15328`)
 
 
-
+- Bug in ``groupby.transform`` where calculating a ``timedelta`` from a ``datetime`` caused a ``ValueError`` (:issue:`15429`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: just do a single note with the issue number. make this simpler, something like: .transform incorrectly changing the dtype of the output.

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -2906,8 +2906,15 @@ def transform(self, func, *args, **kwargs):
                 common_type = np.common_type(np.array(res), result)
                 if common_type != result.dtype:
                     result = result.astype(common_type)
-            except:
-                pass
+            except Exception as exc:
```
Comment (jreback) [inline pandas/core/groupby.py]: huh? this is way complicated

**Comment by jorisvandenbossche** [comment]
>  I looked at it, but didn't undertsand it well enough to know how to fix.

@stephenrauch Which test broke when you used `_find_common_type`? Can you try to describe the problem a bit more in detail? 
(also, if you add new commits to a PR instead of amending it to the first, it is easier to see what you changed and to comment on those steps)

**Comment by jreback** [comment]
@stephenrauch I have a fix for this which avoids all of this complication. I will push soon.

File: pandas/core/groupby.py (hunk 3 above)
Comment (stephenrauch) [inline pandas/core/groupby.py]: Sorry, this was a line I put into to see why the change to _find_common_type was causing problems.

**Comment by jreback** [comment]
@stephenrauch ok (aside from a flake issue) I think this should work. But if you'd run the perf suite for groupby to see that wasn't changed too much. IIRC that's the reason we did the inplace indexing; and I think that was before we did the fast_transform anyhow (which is now a different path), so this maybe be ok.

**Comment by jreback** [comment]
thanks @stephenrauch I merged this (I actually needed some more fixes for making sure ``.agg`` didn't do the same thing). And perf testing was about the same.
